### PR TITLE
refactor: ExitSwapExternAmountOut

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #1130

## Description

- Re-implemented `ExitSwapExternAmountOut` with similar logic to v7
- introduced `PoolExitSwapExternAmountOutExtension`

## Risks

I was confused about the following part:
https://github.com/osmosis-labs/osmosis/blob/523ed353f033e8a7c0e9f86c72e6400c19918980/x/gamm/keeper/pool_service.go#L533-L543

I think that instead of doing that, it is possible to call `k.applyExitPoolStateChange(...)` so that it takes care of the fee since it does `k.BurnPoolShareFromAccount(ctx, pool, exiter, numShares)` where `numShares` do not have the `exitFee` subtracted. 

It seems to me that in the old logic subtracting the fee only to burn it separately was redundant, and I'm not sure if I'm missing some algebra reasoning behind it.

## Next Steps

I'm planning to spend some time writing more unit tests for this in a separate PR. Then, spearhead more e2e tests by joining a swap and exiting it using this method. 

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: x/gamm/types/pool.go [hunk 1]
```diff
@@ -74,6 +74,15 @@ type PoolI interface {
 	PokePool(blockTime time.Time)
 }
 
+type PoolExitSwapExternAmountOutExtension interface {
```
Comment (p0mvn) [inline x/gamm/types/pool.go:77]: TODO: godoc

Comment (p0mvn) [inline x/gamm/types/pool.go:77]: TODO: godoc

File: CHANGELOG.md [hunk 2]
```diff
@@ -43,6 +43,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Features
 
+* [#1244](https://github.com/osmosis-labs/osmosis/pull/1244) refactor: ExitSwapExternAmountOut
```
Comment (alexanderbez) [inline CHANGELOG.md:46]:
```suggestion
* [#1244](https://github.com/osmosis-labs/osmosis/pull/1244) Refactor `x/gamm`'s `ExitSwapExternAmountOut`.
```

File: x/gamm/types/pool.go (hunk 1 above)
Comment (alexanderbez) [reply x/gamm/types/pool.go:77]: bump

File: x/gamm/keeper/pool_service.go [hunk 3]
```diff
@@ -401,13 +401,20 @@ func (k Keeper) ExitSwapExternAmountOut(
...
+	if !ok {
+		return sdk.Int{}, fmt.Errorf("pool with id %d does not support this kind of exit", poolId)
+	}
+
+	shareInAmount, err = extendedPool.ExitSwapExternAmountOut(ctx, tokenOut, shareInMaxAmount)
+
+	if err := k.applyExitPoolStateChange(ctx, pool, sender, shareInAmount, sdk.Coins{tokenOut}); err != nil {
+		return sdk.Int{}, err
+	}
+	return shareInAmount, nil
```
Comment (alexanderbez) [inline x/gamm/keeper/pool_service.go:419]:
```suggestion

	return shareInAmount, nil
```

File: x/gamm/pool-models/balancer/amm.go [hunk 4]
```diff
@@ -375,3 +375,62 @@ func (p *Pool) CalcExitPoolShares(ctx sdk.Context, exitingShares sdk.Int, exitFe
 	}
 	return exitedCoins, nil
 }
+
+// balancer notation: pAi - poolshares amount in, given single out
```
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/amm.go:379]:
```suggestion
// balancer notation: pAi - pool shares amount in, given single out
```

Comment (alexanderbez) [reply x/gamm/pool-models/balancer/amm.go:379]: What is "single out"? I'm not sure how to read this godoc :P

File: x/gamm/pool-models/balancer/amm.go [hunk 5]
```diff
@@ -375,3 +375,62 @@ func (p *Pool) CalcExitPoolShares(ctx sdk.Context, exitingShares sdk.Int, exitFe
...
+// the second argument requires the tokenWeightOut / total token weight.
+func calcPoolInGivenSingleOut(
+	tokenBalanceOut,
+	normalizedTokenWeightOut,
+	poolSupply,
+	tokenAmountOut,
+	swapFee sdk.Dec,
```
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/amm.go:386]:
```suggestion
	swapFee,
```

File: x/gamm/pool-models/balancer/amm.go [hunk 6]
```diff
@@ -375,3 +375,62 @@ func (p *Pool) CalcExitPoolShares(ctx sdk.Context, exitingShares sdk.Int, exitFe
...
+	exitFee sdk.Dec,
+) sdk.Dec {
+	feeRatio := sdk.OneDec().Sub((sdk.OneDec().Sub(normalizedTokenWeightOut)).Mul(swapFee))
```
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/amm.go:389]:
```suggestion
	// feeRatio is defined as follows:
	// 1 - ((1 - normalizedTokenWeightOut) * swapFee)
	feeRatio := sdk.OneDec().Sub((sdk.OneDec().Sub(normalizedTokenWeightOut)).Mul(swapFee))
```

File: x/gamm/pool-models/balancer/amm.go [hunk 7]
```diff
@@ -375,3 +375,62 @@ func (p *Pool) CalcExitPoolShares(ctx sdk.Context, exitingShares sdk.Int, exitFe
...
+		return sdk.Int{}, err
+	}
+
+	poolAmountInBeforeFee := calcPoolInGivenSingleOut(
+		pAsset.Token.Amount.ToDec(),
+		pAsset.Weight.ToDec().Quo(p.TotalWeight.ToDec()),
+		p.GetTotalShares().ToDec(),
+		tokenOut.Amount.ToDec(),
+		p.GetSwapFee(ctx),
+		p.GetExitFee(ctx))
```
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/amm.go:419]:
```suggestion
		p.GetExitFee(ctx),
	)
```

File: x/gamm/pool-models/balancer/amm.go [hunk 8]
```diff
@@ -375,3 +375,62 @@ func (p *Pool) CalcExitPoolShares(ctx sdk.Context, exitingShares sdk.Int, exitFe
...
+
+	if poolAmountInBeforeFee.LTE(sdk.ZeroInt().ToDec()) {
```
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/amm.go:421]: Pretty sure there is a `sdk.ZeroDec` that we can use.

File: x/gamm/pool-models/balancer/amm.go [hunk 9]
```diff
@@ -375,3 +375,62 @@ func (p *Pool) CalcExitPoolShares(ctx sdk.Context, exitingShares sdk.Int, exitFe
...
+		return sdk.Int{}, sdkerrors.Wrapf(types.ErrInvalidMathApprox, "token amount is zero or negative")
+	}
+
+	if poolAmountInBeforeFee.GT(shareInMaxAmount.ToDec()) {
```
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/amm.go:425]: I'm not super familiar with the balancer amm arithmetic yet, but why do we compare balances to shares here?

File: x/gamm/keeper/pool_service.go [hunk 10]
```diff
@@ -401,13 +401,20 @@ func (k Keeper) ExitSwapExternAmountOut(
...
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	extendedPool, ok := pool.(types.PoolExitSwapExternAmountOutExtension)
+	if !ok {
+		return sdk.Int{}, fmt.Errorf("pool with id %d does not support this kind of exit", poolId)
+	}
+
+	shareInAmount, err = extendedPool.ExitSwapExternAmountOut(ctx, tokenOut, shareInMaxAmount)
```
Comment (ValarDragon) [inline x/gamm/keeper/pool_service.go:414]: This error needs to be checked

File: x/gamm/pool-models/balancer/balancer_pool.go [hunk 11]
```diff
@@ -13,7 +13,7 @@ import (
 	"github.com/osmosis-labs/osmosis/v7/x/gamm/types"
 )
 
-var _ types.PoolI = &Pool{}
+var _ types.PoolExitSwapExternAmountOutExtension = &Pool{}
```
Comment (ValarDragon) [inline x/gamm/pool-models/balancer/balancer_pool.go:16]: Can we actually do a separate such interface assertion line for every expected interface. (I know that `PoolExitSwapExternAmountOutExtension` implies PoolI, but I think these are often to simplify things for the reader)

File: x/gamm/pool-models/balancer/amm.go (hunk 7 above)
Comment (p0mvn) [reply x/gamm/pool-models/balancer/amm.go:419]: changed to:
```
		p.GetExitFee(ctx),
	)
```

File: x/gamm/pool-models/balancer/amm.go (hunk 9 above)
Comment (ValarDragon) [reply x/gamm/pool-models/balancer/amm.go:428]: ah we should rename `calcPoolOutGivenSingleIn` to `calcPoolSharesOutGivenSingleIn` and the same for the variable, then this makes sense / is correct.

File: x/gamm/pool-models/balancer/amm.go [hunk 12]
```diff
@@ -375,3 +375,65 @@ func (p *Pool) CalcExitPoolShares(ctx sdk.Context, exitingShares sdk.Int, exitFe
...
+		return sdk.Int{}, sdkerrors.Wrapf(types.ErrLimitMaxAmount, "%s token is larger than max amount", pAsset.Token.Denom)
+	}
+
+	pAsset.Token.Amount = pAsset.Token.Amount.Sub(tokenOut.Amount)
+	err = p.UpdatePoolAssetBalance(pAsset.Token)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
```
Comment (ValarDragon) [inline x/gamm/pool-models/balancer/amm.go:437]: We need to update the total shares as well, ala

https://github.com/osmosis-labs/osmosis/pull/1244/files#diff-897e45c36167c79b080379402216acbe7c118711dcaa23ca7852cd2a70aca3c1R346

Maybe we refactor share updating + token balance updating for exits into a single re-used function between the two?

File: x/gamm/pool-models/balancer/amm.go (hunk 9 above)
Comment (p0mvn) [reply x/gamm/pool-models/balancer/amm.go:428]: Although renaming calcPoolOutGivenSingleIn to calcPoolSharesOutGivenSingleIn makes sebse, I'm assuming that you are referring to renaming calcPool**In**GivenSingle**Out** to calcPoolShares**In**GivenSingleAsset**Out** in this particular example of `ExitSwapExternAmountOut`?

File: x/gamm/pool-models/balancer/amm.go (hunk 12 above)
Comment (p0mvn) [reply x/gamm/pool-models/balancer/amm.go:445]: Good idea, I extracted this functionality from `ExitPool` into `exitPool`

File: x/gamm/keeper/pool_service.go [hunk 13]
```diff
@@ -395,20 +395,32 @@ func (k Keeper) ExitSwapShareAmountIn(
...
+	shareInAmount, err = extendedPool.ExitSwapExactAmountOut(ctx, tokenOut, shareInMaxAmount)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	if err := k.applyExitPoolStateChange(ctx, pool, sender, shareInAmount, sdk.Coins{tokenOut}); err != nil {
+		return sdk.Int{}, err
+	}
+
+
```
Comment (alexanderbez) [inline x/gamm/keeper/pool_service.go:424]:
```suggestion
```

File: x/gamm/pool-models/balancer/amm.go [hunk 14]
```diff
@@ -330,22 +330,31 @@ func (p *Pool) CalcJoinPoolShares(_ctx sdk.Context, tokensIn sdk.Coins, swapFee
...
-	if err != nil {
+	if err := p.exitPool(ctx, exitingCoins, exitingShares); err != nil {
 		return sdk.Coins{}, err
 	}
 
+	return exitingCoins, nil
+}
+
+// exitPool exits the pool given exitingCoins and exitingShares.
+// updates the pool's liquidity and totalShares
```
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/amm.go:347]:
```suggestion
// updates the pool's liquidity and totalShares.
```

**Review by alexanderbez — APPROVED** [review]

File: x/gamm/pool-models/balancer/amm.go (hunk 12 above)
Comment (ValarDragon) [reply x/gamm/pool-models/balancer/amm.go:445]: Nice! That change LGTM

**Review by ValarDragon — APPROVED** [review]

**Pull request merged**